draft-irtf-cfrg-opaque.md

```diff
@@ -954,11 +955,89 @@ The server state may have the following fields:
 - server_ake_state: The server's AKE state if necessary.
 
 The rest of this section describes these authenticated key exchange messages
-and their parameters in more detail. {{cred-retrieval}} discusses internal
-functions used for retrieving client credentials, and {{ake-protocol}} discusses
-how these functions are used to execute the authenticated key exchange protocol.
+and their parameters in more detail. {{ake-messages}} defines the structure of the
+messages passed between client and server in the above setup. {{ake-functions}}
+describes details of the functions and corresponding parameters mentioned above.
+{{cred-retrieval}} discusses internal functions used for retrieving client
+credentials, and {{ake-protocol}} discusses how these functions are used to execute
+the authenticated key exchange protocol.
 
-## Client Authentication Functions {#opaque-client}
+## AKE Messages {#ake-messages}
+
+In this section, we define the `KE1`, `KE2`, and `KE3` structs which make up
+the AKE messages used in the protocol. `KE1` is composed of a `CredentialRequest`
+and `AuthRequest`, and `KE2` is composed of a `CredentialResponse`
+and `AuthResponse`.
+
+~~~
+struct {
+  uint8 client_nonce[Nn];
+  uint8 client_keyshare[Npk];
+} AuthRequest;
+~~~
+
+client_nonce: A fresh randomly generated nonce of length `Nn`.
+
+client_keyshare: A serialized client ephemeral key share of fixed size `Npk`.
+
+~~~
+struct {
+  CredentialRequest credential_request;
+  AuthRequest auth_request;
+} KE1;
+~~~
+
+credential_request: A `CredentialRequest` structure.
+
+auth_request: An `AuthRequest` structure.
+
+~~~
+struct {
+  uint8 server_nonce[Nn];
+  uint8 server_keyshare[Npk];
+  uint8 server_mac[Nm];
+} AuthResponse;
+~~~
+
+server_nonce: A fresh randomly generated nonce of length `Nn`.
+
+server_keyshare: Server ephemeral key share of fixed size `Npk`, where `Npk`
+depends on the corresponding prime order group.
+
+server_mac: An authentication tag computed over the handshake transcript
+computed using `Km2`, defined below.
+
+~~~
+struct {
+  CredentialResponse credential_response;
+  AuthResponse auth_response;
+} KE2;
+~~~
+
+credential_response: A `CredentialResponse` structure.
+
+auth_response: An `AuthResponse` structure.
+
+~~~
+struct {
+  uint8 client_mac[Nm];
+} KE3;
+~~~
+
+client_mac: An authentication tag computed over the handshake transcript
+of fixed size `Nm`, computed using `Km2`, defined below.
+
+## AKE Functions {#ake-functions}
+
+In this section, we define the main functions used to produce the AKE messages
+in the protocol. Note that this section relies on definitions of subroutines defined
+in later sections:
+- `CreateCredentialRequest`, `CreateCredentialResponse`, `RecoverCredentials`
+  defined in {{cred-retrieval}}
+- `AuthStart`, `AuthRespond`, `AuthClientFinalize`, and `AuthServerFinalize`
```

Comment:
Should we make `AuthStart` be `AuthClientStart` (and similar for AuthRespond) so the roles are explicit?

Reply:
Good idea! Updated in the latest commit.